### PR TITLE
crimson/osd: Support snapshot rollbacks 

### PR DESCRIPTION
**Pull request by Matan-B**
```
test_rados.TestIoCtxSelfManagedSnaps.test ... ok     
test_rados.TestIoctx.test_snap_rollback ... ok 
test_rados.TestIoctx.test_snap_rollback_removed ... ok
                                          
----------------------------------------------------------------------
Ran 3 tests in 19.729s 
                                            
OK 
```
```
$ CRIMSON_COMPAT=true RBD_FEATURES=1 bin/ceph_test_librbd --gtest_filter=TestLibRBD.TestIOToSnapshotseed 480475                        
Note: Google Test filter = TestLibRBD.TestIOToSnapshot                                                                                                                                        
[==========] Running 1 test from 1 test suite.                                                                                                                                                
[----------] Global test environment set-up.                                                                                                                                                  
[----------] 1 test from TestLibRBD                                                                                                                                                           
[ RUN      ] TestLibRBD.TestIOToSnapshot
[       OK ] TestLibRBD.TestIOToSnapshot (7290 ms)
[----------] 1 test from TestLibRBD (7290 ms total)
...
[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (8217 ms total)
[  PASSED  ] 1 test.

```

~Blocked by: #48682~

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by Matan-B** [comment]
jenkins test make check

File: src/crimson/osd/pg_backend.cc [hunk 1]
```diff
@@ -766,53 +767,52 @@ PGBackend::write_iertr::future<> PGBackend::writefull(
...
+    os.oi.clear_flag(object_info_t::FLAG_WHITEOUT);
+    os.oi.copy_user_bits(clone_obc->obs.oi);
+    // TODO: 3) Calculate clone_overlaps by following overlaps
+    //          forward from rollback snapshot
+    return rollback_iertr::now();
+  }).handle_error_interruptible(
+    crimson::ct_error::enoent::handle([this, &os, &snapid] {
+      logger().error("PGBackend::rollback: {} clone object"
+                     " with snap_id of {} not found",
+                     os.oi.soid, snapid);
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:808]: Isn't this case supposed to be a delete?

File: src/crimson/osd/pg_backend.cc [hunk 2]
```diff
@@ -766,53 +767,52 @@ PGBackend::write_iertr::future<> PGBackend::writefull(
...
+    txn.remove(coll->get_cid(), ghobject_t{os.oi.soid,
+                                ghobject_t::NO_GEN, shard});
+    // 2) Clone correct snapshot into head
+    txn.clone(coll->get_cid(), ghobject_t{clone_obc->obs.oi.soid},
+                               ghobject_t{os.oi.soid});
+    //    Copy clone obc.os.oi to os.oi
+    os.oi.clear_flag(object_info_t::FLAG_WHITEOUT);
+    os.oi.copy_user_bits(clone_obc->obs.oi);
+    // TODO: 3) Calculate clone_overlaps by following overlaps
+    //          forward from rollback snapshot
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:802]: Seems pretty straightforward to just add this code.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/pg_backend.cc (hunk 2 above)
Comment (Matan-B) [reply src/crimson/osd/pg_backend.cc:805]: 625d8102156438e581799822bce74f9e44171bbd

File: src/crimson/osd/pg_backend.cc (hunk 1 above)
Comment (Matan-B) [reply src/crimson/osd/pg_backend.cc]: Please see c5218903239564dea272b63ab4ec2e0fc696836a.
I couldn't use `remove()` with `handle_error_interruptible` constrains.
One way to solve that was using the `delete_head` flag, do you find that acceptable?

File: src/crimson/osd/pg_backend.cc [hunk 3]
```diff
@@ -766,53 +767,68 @@ PGBackend::write_iertr::future<> PGBackend::writefull(
...
+    //          https://tracker.ceph.com/issues/58263
+    return rollback_iertr::now();
+  }).handle_error_interruptible(
+    crimson::ct_error::enoent::handle(
+      [this, &os, &snapid, &txn, &delta_stats, &delete_head] {
+      // there's no snapshot here, or there's no object.
+      // if there's no snapshot, we delete the object;
+      // otherwise, do nothing.
+      delete_head = true;
+      return rollback_ertr::now();
```
Comment (rzarzynski) [inline src/crimson/osd/pg_backend.cc:815]: Why not `seastar::now()` (or an unerrorated but interruptible variant of it)?

Why exactly we can't do the `remove()` logic here? Dividing into two phases seems a bit problematic (e.g. needs ensuring proper life-time for the `delete_head` flag).

File: src/crimson/osd/pg_backend.cc [hunk 4]
```diff
@@ -766,53 +767,68 @@ PGBackend::write_iertr::future<> PGBackend::writefull(
...
-                   " because got ENOENT|whiteout on obc lookup",
-                   os.oi.soid);
-    return remove(os, txn, delta_stats, true /*whiteout*/);
-  }), mocked_load_clone_obc_ertr::assert_all{
-    "unexpected error code in rollback"
-  });
+                  __func__, os.oi.soid ,snapid);
+  hobject_t target_coid = os.oi.soid;
+  target_coid.snap = snapid;
+  bool delete_head = false;
```
Comment (rzarzynski) [inline src/crimson/osd/pg_backend.cc:786]: On stack but taken a by reference in the lambdas.

File: src/crimson/osd/pg_backend.cc (hunk 3 above)
Comment (Matan-B) [reply src/crimson/osd/pg_backend.cc:815]: I strongly agree it seems problematic, please see my previous comment to Sam.
(Note, I did not squash the commit yet and left it as DNM for easier discussion)

Comment (rzarzynski) [reply src/crimson/osd/pg_backend.cc:815]: I saw only this one:

> I couldn't use `remove()` with `handle_error_interruptible` constrains.

but it doesn't _exactly_ explain what was the issue. I guess `remove()` returns an errorated-and-interruptible future and I speculate that, in the handler above, you want to return pure interruptible one. If that's the case, you might need another, inner `handle_error_interruptible` to handle (`assert_all{}`?) errors just from `remove()`.

Comment (Matan-B) [reply src/crimson/osd/pg_backend.cc:815]:
```
return obc_loader.with_clone_obc_only<...>(...) {
...
}).handle_error_interruptible(
  crimson::ct_error::enoent::handle(
    [this, &os, &snapid, &txn, &delta_stats, &delete_head] {
    return remove(os, txn, delta_stats, false).handle_error_interruptible(
      crimson::ct_error::assert_all{});
  }),
  crimson::osd::ObjectContextLoader::load_obc_ertr::all_same_way(
  [target_coid](auto& code) {
    ...
}));
```

`handle_error_interruptible` expects to return `errorator`:
```
../src/crimson/osd/pg_backend.cc:808:32:   required from here
../src/crimson/common/errorator.h:339:88: error: no type named ‘errorator_type’ in ‘using type = using type = class seastar::future<>’ {aka ‘class seastar::future<>’} 
```
(l.808 is `}).handle_error_interruptible(`) mentioned here.

Comment (Matan-B) [reply src/crimson/osd/pg_backend.cc]: I ended up handling `enoent` and `assert_all` errors separately by `safe_then_interruptible`.

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
Merging based on:
https://pulpito.ceph.com/matan-2023-01-11_14:07:51-crimson-rados-wip-matanb-crimson-testing-2.10-v2-distro-crimson-smithi/

**Pull request merged**